docs/API.md

```diff
@@ -0,0 +1,267 @@
+
+An example in which we override the authentication handling:
+
+```
+App.initialize({
+  overrideHandlers: {
+    authentication: () => {
+      // As a usage example of overriding one phase of the startup sequence,
+      // providing this function will override the default authentication
+      // initialization.
+    }
+  }
+});
+```
+
+`overrideHandlers` has keys corresponding to the eight lifecycle events.  Including a key will override and replace the corresponding lifecycle handler if provided.  Please see Initialization Lifecycle Phases for more information on the phases responsibilities.
+
+#### custom
+
+The `custom` property can be used to attach custom data to the `App` which will be exposed at `App.custom`.  This data can be used in custom initialization handlers, or elsewhere in the application as necessary.  Note, if you're using this to provide mutable data to the application, _strongly_ consider using React props, context, or Redux instead.
+
+### `App.config`
+
+The environment configuration.  Contains the following keys:
+
+- ACCESS_TOKEN_COOKIE_NAME
+- BASE_URL
+- CREDENTIALS_BASE_URL
+- CSRF_COOKIE_NAME
+- CSRF_TOKEN_API_PATH
+- ECOMMERCE_BASE_URL
+- ENVIRONMENT
+- LANGUAGE_PREFERENCE_COOKIE_NAME
+- LMS_BASE_URL
+- LOGIN_URL
+- LOGOUT_URL
+- MARKETING_SITE_BASE_URL
+- ORDER_HISTORY_URL
+- REFRESH_ACCESS_TOKEN_ENDPOINT
+- SECURE_COOKIES
+- SEGMENT_KEY
+- SITE_NAME
+- USER_INFO_COOKIE_NAME
+
+If additional, dynamic config is desired, it would be reasonable to add those keys into `App.config`.
+
+### `App.apiClient`
+
+A reference to the @edx/frontend-auth authenticated API Client.
+
+### `App.subscribe(type, callback)`
+
+A method allowing consumers of `App` to subscribe to lifecycle events.  `type` is an event type, as documented in "Initialization Lifecycle Phases".  There are constants for all the event types:
+
+```
+import {
+  APP_BEFORE_INIT, APP_CONFIGURED, APP_AUTHENTICATED, APP_I18N_CONFIGURED, APP_LOGGING_CONFIGURED, APP_ANALYTICS_CONFIGURED, APP_BEFORE_READY, APP_READY, APP_ERROR
+} from `@edx/frontend-base`
+```
+
+### `App.requireConfig(keys, requester)`
+
+A method allowing application code to indicate that particular `App.config` keys are required for them to function.  The method returns the required config values.
+
+```
+const config = App.requireConfig(['LMS_BASE_URL', 'LOGIN_URL'], 'The name of my feature/component');
+```
+
+**NOTE**: If you use App.requireConfig to require config that hasn't been loaded yet (i.e., from a custom `configuration` handler), it will throw an error.  If you use a custom `configuration` handler, you can defer requiring your config by subscribing to the `APP_CONFIGURED` event:
+
+```
+let config = null;
+App.subscribe(APP_CONFIGURED, () => {
+  config = App.requireConfig(['DYNAMICALLY_CONFIGURED_URL'], 'Consumer of custom config');
```

Comment:
Does this mean that validating config is a later lifecycle that is just unpublished?  You could consider documenting parts of the lifecycle that don't have an event, if events aren't needed.

Reply:
Yeah, this was a bit confusing... I've tried to clarify that `App.config` is generally available _immediately_, but if you add additional configuration in during the configuration phase, you can't access it until that phase is complete.